src/cli/vector_generate.rs

```diff
@@ -0,0 +1,300 @@
+        for file in f {
+            let contents = match fs::read_to_string(file.clone()) {
+                Ok(c) => c,
+                Err(e) => {
+                    return Err(ShellError::GenericError(
+                        format!("Error parsing file {}: {}", file, e),
+                        "".to_string(),
+                        None,
+                        None,
+                        Vec::new(),
+                    ))
+                }
+            };
+
+            let file_chunks = &mut chunk_text(contents, chunk_len);
+            chunks.append(file_chunks);
+        }
+    }
+
+    let bpe = p50k_base().unwrap();
+    let tokens = bpe.encode_with_special_tokens(&chunks.join(" "));
+
+    debug!("Total tokens: {:?}\n", tokens.len());
+
+    let num_batches = (tokens.len() / max_tokens) + 1;
+    let batch_size = chunks.len() / num_batches;
+    if num_batches == 1 {
+        batches.push(chunks.to_vec());
+    } else {
+        let mut lower = 0;
+        let mut upper = batch_size;
+        while lower < chunks.len() {
+            batches.push(chunks[lower..=upper].to_vec());
+            lower = upper + 1;
+            upper += batch_size;
+
+            if upper >= chunks.len() {
+                upper = chunks.len() - 1;
+            }
+        }
+    };
+
+    let client =
+        Client::with_config(async_openai::config::OpenAIConfig::default().with_api_key(key));
+
+    let mut vector_count = 0;
+    let start = SystemTime::now();
+    let rt = Runtime::new().unwrap();
+    for (i, batch) in batches.iter().enumerate() {
+        let batch_start = SystemTime::now();
+        info!("Embedding batch {:?}/{} ", i + 1, batches.len());
+
+        if log::log_enabled!(log::Level::Debug) {
+            let bpe = p50k_base().unwrap();
+            let tokens = bpe.encode_with_special_tokens(&batch.join(" "));
+            debug!("- Tokens: {:?}", tokens.len());
+        }
+
+        let request = CreateEmbeddingRequestArgs::default()
+            .model("text-embedding-3-small")
+            .dimensions(dim)
+            .input(batch.clone())
+            .build()
+            .unwrap();
+
+        let embeddings = client.embeddings();
+        let ctrl_c = engine_state.ctrlc.as_ref().unwrap().clone();
+        let ctrl_c_fut = CtrlcFuture::new(ctrl_c);
+        let response = match rt.block_on(async {
+            select! {
+                result = embeddings.create(request) => {
+                    match result {
+                        Ok(r) => Ok(r),
+                        Err(e) => Err(ShellError::GenericError(
+                            format!("failed to execute request: {}", e),
+                            "".to_string(),
+                            None,
+                            None,
+                            Vec::new(),
+                        ))
+                    }
+                },
+                () = ctrl_c_fut =>
+                     Err(ShellError::GenericError(
```

Comment:
Is there a way to cancel the call via openai if ctrlc is pressed?

Reply:
Not that is immediately obvious, I will make a note of this as an improvement but don't want to block this merge for that.